core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -729,10 +729,9 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     * }}}
     */
   def map[O2](f: O => O2): Stream[F, O2] =
-    this.repeatPull(_.uncons.flatMap {
-      case None           => Pull.pure(None);
-      case Some((hd, tl)) => Pull.output(hd.map(f)).as(Some(tl))
-    })
+    mapSegments(_.map(f))
+  // TODO replace the above with the following once ResourceSafetySpec passes
```

Comment:
@pchlupacek @AdamChlupacek Could you take a look at this? If I replace the definition of `map` with the commented out version, the `finalizers are run in LIFO order - scope closure` test fails due to the innermost resource not being released first but rather last. That is, the ordering of finalizers ends up 1, 2, 3, ..., 8, 9, 0 instead of the correct 0, 1, 2, 3, ... 8, 9. I added a bunch of logging and noticed the following difference: with the old definition of `map`, once the exception is thrown, the scope is closed, resulting in all 10 resources being released. With the new definition, once the exception is thrown, the 1st resource is released, then the 2nd, then the 3rd, and so on up to and including the 9th -- then the scope is closed, resulting in the outstanding resource 0 getting finalized. Hence, it appears we are somehow skipping some error handling?

Reply:
Ok, So the issue here is that `Segment` is actually lazy. As such we do not know about the error until we force the segment. 

This is happening in `compileScope`. Where once we encounter error in the segment we actually use the `tail` as the error handling. Thus we skip the error handling for the `emit` of `0` and go  straight to the error handling of `1`. 

The difference between this and how it works with `mapSegments`, is the existence of `uncons` which is now defined using `Step`, as the `Step` has the tail of the `stream` in it, it is not consulted and the scope is directly closed thus we get the correct order of resources released. 

A possible fix is that we will add the error handler of the of the `Segment` to the `Out` and then inside of the `compileScope` we use this error handler instead of `tail.asHandler`.